lib/matplotlib/image.py

```diff
@@ -361,6 +362,13 @@ def _make_image(self, A, in_bbox, out_bbox, clip_bbox, magnification=1.0,
                     a_min, a_max = np.int32(0), np.int32(1)
                 if inp_dtype.kind == 'f':
                     scaled_dtype = A.dtype
+                    # Cast to float64
+                    if A.dtype not in (np.float32, np.float16):
+                        if A.dtype != np.float64:
```

Comment:
might as well just make this one if-statement....

Reply:
There are things happening after this one but not in it...

Reply:
```python
if A.dtype not in (np.float32, np.float16, np.float64):
   warnings.warn()
```
should do it, right?

Reply:
I actually did this in commit a453ba4. This led to problems in some Windows environments and failures in CI. It seems like scaled_dtype needs to be set to np.float64 instead of np.longdouble even if np.longdouble == np.float64.

Reply:
`not in` returns something different than `!=`?  I'm not following the logic here at all.

Reply:
I see now. Yes, you are correct. I forgot that in (...) is not using an equality check. This should make things much simpler.

Reply:
The extra if statement is because on Windows environment, np.doublelong == float64 but np.longdouble is not float64. We need to make dtype to be actually float64 and not just dtype == float64 to return True.

Reply:
Gotchya, so https://github.com/matplotlib/matplotlib/commit/a453ba4519d66e3ba2dd8bab3d64294ba34258cc failed because `np.doublelong == float64 ` but this should work because `np.doublelong is not in (np.float64)` .  How confusing!  Whatever you decide, this definitely needs a comment so its clear to future readers what is going on.  Sounds like you have it sorted out, but its pretty arcane.

Reply:
I think what you're missing is the [next line](https://github.com/matplotlib/matplotlib/pull/10768/files#diff-d7067d23b8918c423e30d4a3c6e13a60R371). The difference has nothing to do with `==` vs `in`, but that `scaled_dtype` is correctly updated (which causes some problem later).

Reply:
If `A.dtype` is `in (np.float64)` then `scaled_dtype` is already `np.float64` per line 364 above, I'm 95% sure.

Reply:
It could be `np.longdouble` as that also equals `np.float64` (on some systems). Why this causes a problem later, I don't know...

Reply:
I believe somewhere later, it checks that the `scaled_dtype` is `np.float32`, `np.float64`, etc. instead of ==. That's why `scaled_dtype` needs to be set to `np.float64`.

Reply:
@br-Zhang I think @QuLogic is correct.  On my machine `np.longdouble` is `np.float128` and:

```
In [7]: print(np.longdouble == np.float128)
True

In [8]: print(np.longdouble in (np.float64, np.float32, np.float128))
True
```

so I don't understand at all whats going on.

Reply:
It's very platform specific. Running on Win-64 gives:
`
>>> numpy.longdouble in (numpy.float16, numpy.float64)
False
`

Reply:
But `numpy.longdouble == numpy.float64` returns true?

Reply:
Very weird.
``` 
>>> numpy.longdouble == numpy.float64
True
```

Reply:
Again, the problem doesn't have anything to do with the check here; it's the use of `scaled_dtype` later that causes a problem. If you flatten the condition, then `scaled_dtype=np.longdouble` is not replaced by `np.float64` (on Windows only!) It's then crashing in C code later (which is arguably a NumPy inconsistency) which doesn't understand `np.longdouble`.

Reply:
OK; I get it now.  everything not `np.float16/32` has to be explicitly set to `np.float64`, even if python thinks `np.longdouble` is equivalent to `np.float64`, because the C code doesn't think its equivalent.  

Sorry to be so dense, but I do think its pretty bad that the C code ends up incompatible in this way.